README.md
<!--
 * @Author: “wuziyang” “764401855@qq.com”
 * @Date: 2023-03-31 11:32:54
 * @LastEditors: “wuziyang” “764401855@qq.com”
 * @LastEditTime: 2023-04-19 21:04:14
 * @FilePath: \pitsftp\README.md
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
-->
## pitSftp插件
作者：湖南创智艾泰克-前端组
#### 介绍
vscode插件，上传文件至服务器，密码加密
#### 使用教程

1.  安装插件之后，我们就可以进行密码加密，ctrl+alt+p 弹出输入框，输入服务器密码，回车您将获得一个经过非对称加密的密码，并且在项目根目录的.vscode文件夹下生成一个pitSftp.json的配置文件
``` json
{
    "name": "My Server", // 配置文件的名字
    "host": "127.0.0.0", // 服务器的ip
    "protocol": "sftp", // 协议，一般不需要修改
    "port": 22,  // 端口号
    "username": "admin", // 服务器的用户名
    "remotePath": "/etc/nginx/html", // 项目部署的地址
    "key": "24e00495-5493-4b7e-9f99-e31c3fabcf75", // 这个key一般不需要修改，自动生成，用于识别公钥与私钥的存放路径
    "password": "GMzmnJrMAG8dWNqwOjpD6aSxUUKdhOnoOrnUIoK/flo9IodWp8u2GWx42BUE0OZ6dU1FTDvkuGF3Hf2ji3yuhw==" // 加密后的密码
}

```
2.  选择要上传的文件夹，右键选择PITSFTP:上传至目标服务器
![Image text](./img/youjian.png)
3.  耐心等待一下，注意右下角弹出的提示

#### 使用说明



#### 参与贡献

1.  Fork 本仓库
2.  新建 Feat_xxx 分支
3.  提交代码
4.  新建 Pull Request



CHANGELOG.md
# Change Log

All notable changes to the "pitsftp" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

sftp.js
/*
 * @Author: “wuziyang” “764401855@qq.com”
 * @Date: 2023-04-10 16:38:54
 * @LastEditors: “wuziyang” “764401855@qq.com”
 * @LastEditTime: 2023-04-19 20:52:52
 * @FilePath: \pitsftp\sftp.js
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
let Client = require('ssh2-sftp-client');
const vscode = require("vscode");
const NodeRSA = require('node-rsa');
const fs = require("fs")
let path = require('path');
const Progress = require('./utils/Progress')
// const basePath = vscode.workspace.workspaceFolders[0].uri.fsPath;
// const rootPath = path.join(basePath, CONFIG_PATH);
// const shell = require("shelljs");
// const cmdShell = require('node-cmd')

// const localPath = path.resolve(__dirname, 'C://Users//hyh//Desktop//company-project//pit-jxh-web//dist')
// const remotePath = '/data/docker/nginx/html/web'
let client = new Client();
module.exports = async (option, explorer) => {
  const localPath = explorer.path[0] == '/' ? explorer.path.slice(1) : explorer.path
  console.log(localPath, 'localPath')
  const stat = fs.statSync(localPath)
  if (!stat.isDirectory()) {
    return vscode.window.showWarningMessage('目前仅支持上传文件夹')
  }
  const privateKey = fs.readFileSync(path.join(__dirname, `./pem/${option.key}/privateKey.pem`), "utf8")
  const priKey = new NodeRSA(privateKey,'pkcs8-private');
  var decrypted = priKey.decrypt(option.password, 'utf8');
  const { progress } = new Progress('PITSFTP文件传输', vscode.ProgressLocation.Notification)
  let i = 0
  let num = 0
  let files = getAllFileList(localPath)
  function upload(info) {
    i++
    progress.report({ increment: num === parseInt((i / files.length) * 100) ? 0 : 1, message: "正在上传：\r\n" +  info.source});
    num = parseInt((i / files.length) * 100)
  }
  try {
    progress.report({ increment: 0, message: "正在努力链接服务器。。。。" });
    await client.connect({
        host: option.host,
        port: option.port,
        username: option.username,
        password: decrypted
    })
    // if (await client.exists(option.remotePath)) {
    //   await client.rmdir(option.remotePath, true)
    //   console.log('删除成功')
    // }
    client.on('upload', upload);
    let rslt = await client.uploadDir(localPath, option.remotePath);
    // let rslt = await client.fastPut(localPath, option.remotePath, { step: (total_transferred, chunk, total) => {
    // progress.report({ increment: 0, message: "正在上传：" + });
    // } });
    progress.resolve()
    vscode.window.showInformationMessage('上传成功');
    console.log('上传成功')
    client.removeListener('upload', upload)
    client.end()
    return rslt;
  }
  catch (err) {
    console.log(err, 'client')
    vscode.window.showInformationMessage("连接服务器失败:"+err);
  }
  finally {
    client.end();
  }  
}

function getAllFileList(MyUrl) {
  let fileUrlList = []
  const getFileList = (url) => {
    const fileList = fs.readdirSync(url)
    fileList.forEach(file => {
      let fPath = path.join(url, file);
      const stat = fs.statSync(fPath)
      if (stat.isDirectory()) {
        getFileList(fPath)
          //stat 状态中有两个函数一个是stat中有isFile ,isisDirectory等函数进行判断是文件还是文件夹
      }
      else {
        fileUrlList.push(fPath)
      }
    })
  }
  getFileList(MyUrl)
  return fileUrlList
}

utils/Progress.js
/*
 * @Author: “wuziyang” “764401855@qq.com”
 * @Date: 2023-04-19 10:44:46
 * @LastEditors: “wuziyang” “764401855@qq.com”
 * @LastEditTime: 2023-04-19 20:31:13
 * @FilePath: \pitsftp\utils\Progress.js
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
const { resolve } = require('path');
const vscode = require('vscode');

/**
 * 进度条工具类
 */

module.exports = class Progress {
  static name = '';
  progress = '';
  constructor(name, location) {
    this.name = name
    this.create(location)
  }
  create(location) {
    vscode.window.withProgress({
      location: location || vscode.ProgressLocation.Notification,
      title: this.name || '正在加载',
      cancellable: true
    }, (progress) => {
      return new Promise(resolve => {
        progress.resolve = resolve
        this.progress = progress
      })
      // 初始化进度
      // progress.report({ increment: 0 });

      // setTimeout(() => {
      //   progress.report({ increment: 10, message: "在努力。。。." });
      // }, 1000);
    })
  }
}

commands/pitInputPassword.js
/*
 * @Author: “wuziyang” “764401855@qq.com”
 * @Date: 2023-04-12 14:32:15
 * @LastEditors: “wuziyang” “764401855@qq.com”
 * @LastEditTime: 2023-04-18 18:52:51
 * @FilePath: \pitsftp\commands\pitInputPassword.js
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
const vscode = require('vscode');
const path = require('path');
// const fs = require('fs');
const fs = require('fs-extra')
const NodeRSA = require('node-rsa');
const { CONFIG_PATH } = require("../constants");
const basePath = vscode.workspace.workspaceFolders[0].uri.fsPath;
const rootPath = path.join(basePath, CONFIG_PATH);
const { v4 } = require('uuid')
module.exports = vscode.commands.registerCommand('pitsftp.inputPassWord', function () {
  vscode.window.showInputBox(
    { // 这个对象中所有参数都是可选参数
        password:true, // 输入内容是否是密码
        ignoreFocusOut:true, // 默认false，设置为true时鼠标点击别的地方输入框不会消失
        placeHolder:'输入密码进行加密', // 在输入框内的提示信息
        prompt:'服务器密码进行非对称加密', // 在输入框下方的提示信息
        // validateInput:function(text){return text;} // 对输入内容进行验证并返回
    }).then(function(secret){
      if (!secret) return;
      const key = new NodeRSA({b: 512});
      fs.mkdir(path.join(__dirname, '../pem'),err => {
        let option = {}
        fs.pathExists(rootPath).then((exist) => {
          if (exist) {
            fs.readJSON(rootPath, (err1, obj) => {
              if (err1) {
                throw err1
              }
              option = obj
            encryKey(option, key, secret)
          })
          } else {
            vscode.window.showErrorMessage("未找到pitSftp.json文件，自动创建2");
            option = {
              "name": "My Server",
              "host": "localhost",
              "protocol": "sftp",
              "port": 22,
              "username": "username",
              "remotePath": "/",
              "key": v4()
            }
            fs.outputJson(rootPath, option, { spaces: 4 }).then((err) => {
              if (err) throw err
              vscode.window.showTextDocument(vscode.Uri.file(rootPath))
              encryKey(option, key, secret)
            })
          }
        })
      })
  });
});

function encryKey(option, key, secret) {
  const pemMk = path.join(__dirname, `../pem/${option.key}`)
  const publicKeyMkUrl = path.join(pemMk, `./publicKey.pem`)
  const privateKeyMkUrl = path.join(pemMk, `./privateKey.pem`)
  fs.mkdir(pemMk, err => {
    if (err) throw err
    fs.writeFileSync(publicKeyMkUrl, key.exportKey('pkcs8-public')) // 公密
    fs.writeFileSync(privateKeyMkUrl, key.exportKey('pkcs8-private')) // 私密
  })
  const pubKey = new NodeRSA(key.exportKey('pkcs8-public'),'pkcs8-public');
  var encrypt = pubKey.encrypt(Buffer.from(secret), 'base64');
  // const encrypt = privateEncrypt(PRIVKEY, Buffer.from(secret)); // 加密
  // const decrypt = crypto.publicDecrypt(publicKey, encrypt); // 解密
  setPassword(encrypt)
  let options = {
    content: encrypt,
    language:"js"
  };
  
  vscode.workspace.openTextDocument(options).then(doc => {
    vscode.window.showTextDocument(doc, vscode.ViewColumn.One);
  }, err => {
    vscode.window.showErrorMessage(err);
  });
}

function setPassword(passWord) {
  const rootUrl = vscode.workspace.workspaceFolders.length === 1 ? vscode.workspace.workspaceFolders[0].uri.path : null
  if (!rootUrl) return;
  // const localPath = rootUrl[0] == '/' ? rootUrl.slice(1) : rootUrl

  fs.pathExists(rootPath).then(exist => {
    console.log(exist, 'exist')
    if (exist) { // 如果存在
      return fs.readJson(rootPath, (err, data) => {
        if (err) throw err
        data.password = passWord
        setFileText(rootPath, data)
      });
    } else {
      vscode.window.showErrorMessage("未找到pitSftp.json文件，自动创建1");
      setFileText(rootPath, {
        "name": "My Server",
        "host": "localhost",
        "protocol": "sftp",
        "port": 22,
        "username": "username",
        "remotePath": "/",
        "password": passWord,
        "key": v4()
      })
    }
  })

  // rootUrl && fs.readFile(path.join(localPath, CONFIG_PATH), "utf8", function (err, dataStr) {
  //   if (err) {
  //     vscode.window.showErrorMessage("未找到pitSftp.json文件，自动创建");
  //     setFileText(path.join(localPath, CONFIG_PATH), {	
  //       "name": "My Server",
  //       "host": "localhost",
  //       "protocol": "sftp",
  //       "port": 22,
  //       "username": "username",
  //       "password": passWord,
  //       "remotePath": "/",
  //       "key": v4()
  //       }
  //     )
  //   } else {
  //     const option = JSON.parse(dataStr)
  //     option.password = passWord
  //     setFileText(path.join(localPath, CONFIG_PATH), JSON.stringify(option))
  //   }
  // });
}
function setFileText(path, text) {
  fs.outputJson(path, text, { spaces: 4 }).then((err) => {
    if (err) throw err
    vscode.window.showTextDocument(vscode.Uri.file(path))
  })
}

extension.js
/*
 * @Author: “wuziyang” “764401855@qq.com”
 * @Date: 2023-03-31 11:32:54
 * @LastEditors: “wuziyang” “764401855@qq.com”
 * @LastEditTime: 2023-04-19 09:27:47
 * @FilePath: \pitsftp\extension.js
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require("vscode");
const navigation = require("./navigation.js");
const pitInputPassword = require("./commands/pitInputPassword.js");
// const fs = require("fs");
const fse = require('fs-extra')
const { CONFIG_PATH } = require("./constants");
const path = require("path");
const basePath = vscode.workspace.workspaceFolders[0].uri.fsPath;
const rootPath = path.join(basePath, CONFIG_PATH);
const sftpMain = require('./sftp.js')
const { v4 } = require('uuid')
// This method is called when your extension is activated
// Your extension is activated the very first time the command is executed
console.log(vscode, "vscode");
console.log(__filename, "__filename");
// console.log(vscode.workspace.rootPath)
function getFileUri(fileName) {
  return path.join(fileName, rootPath);
}
/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {
  console.log(pitInputPassword)
  context.subscriptions.push(navigation);
  context.subscriptions.push(pitInputPassword);
  let disposable = vscode.commands.registerCommand(
    "pitsftp.upload",
    function (explorer) {
      console.log(explorer, 'explorer')

      fse.pathExists(rootPath).then(exist => {
        if (exist) { // 如果存在
          return fse.readJson(rootPath, (err, data) => {
            if (err) throw err
            sftpMain(data, explorer)
          });
        } else {
          vscode.window.showErrorMessage("未找到pitSftp.json文件，自动创建");
          fse.outputJson(rootPath, {	
            "name": "My Server",
            "host": "localhost",
            "protocol": "sftp",
            "port": 22,
            "username": "username",
            "remotePath": "/",
            "key": v4()
          }, { spaces: 4 }).then((err) => {
            if (err) throw err
            vscode.window.showTextDocument(vscode.Uri.file(rootPath))
          })

        }
      })

//       fse.pathExists(rootPath, "utf8", function (err, dataStr) {
//         console.log(err, "err111");
//         if (err) {
//           vscode.window.showErrorMessage("未找到pitSftp.json文件，自动创建");
//           fs.writeFile(
//             rootPath,
//             `{	
// 	"name": "My Server",
// 	"host": "localhost",
// 	"protocol": "sftp",
// 	"port": 22,
// 	"username": "username",
// 	"remotePath": "/",
// 	"uploadOnSave": false,
// 	"useTempFile": false,
// 	"openSsh": false
// }
// 					`,
//             (error) => {
//               if (!error) vscode.window.showTextDocument(vscode.Uri.file(rootPath));
// 							console.log('error', error)
//             }
//           );
//         } else {
//           sftpMain(JSON.parse(dataStr), explorer)
//           console.log(JSON.parse(dataStr), "dataStr");
//         }
//       });
    }
  );

  context.subscriptions.push(disposable);
}

function deactivate() {}

module.exports = {
  activate,
  deactivate,
};
